discord/voice_client.py

```diff
@@ -293,23 +393,180 @@ def is_connected(self):
 
     # audio related
 
-    def _get_voice_packet(self, data):
-        header = bytearray(12)
-        nonce = bytearray(24)
-        box = nacl.secret.SecretBox(bytes(self.secret_key))
+    @asyncio.coroutine
+    def get_voice_data(self, user, timeout=0.4):
+        """|coro|
 
-        # Formulate header
-        header[0] = 0x80
-        header[1] = 0x78
-        struct.pack_into('>H', header, 2, self.sequence)
-        struct.pack_into('>I', header, 4, self.timestamp)
-        struct.pack_into('>I', header, 8, self.ssrc)
+        Receive voice data from a user until they are silent for
+        `timeout` s.
 
-        # Copy header to nonce's first 12 bytes
-        nonce[:12] = header
+        Parameters
+        -----------
+        user: :class:`abc.User`
+            The user to listen for audio from.
+        timeout: Optional[float]
+            The time in ms of silence after which to stop listening.
+
+        Returns
+        -----------
+        `bytearray`
+            The received voice data from the user.
+
+        Note
+        -----------
+        The returned data is stereo 48kHz audio in raw PCM. You will have to
+        add your own headers if you want to save it as a `.wav` file.
+        """
+        buffer = []
+        last_packet = 0
+        user_id = user.id
+
+        def listener(vc, vp):
+            nonlocal buffer, user_id, last_packet
+            if user_id == (None if vp.user is None else vp.user.id):
+                last_packet = time.time()
+                heapq.heappush(buffer, (vp.timestamp, vp))
```

Comment:
if you're just going to `buffer.sort()` after all this, using a `heapq` is worthless and will just add more work.

Given how python's sorting is implemented, and that packets are probably (for the most part) ordered correctly - you should just call `buffer.sort()` at the end (which should be mostly linear thanks to timsort)